connmgr.go

```diff
@@ -359,9 +366,31 @@ func (cm *BasicConnMgr) getConnsToClose() []network.Conn {
 
 	target := ncandidates - cm.cfg.lowWater
 
-	// slightly overallocate because we may have more than one conns per peer
-	selected := make([]network.Conn, 0, target+10)
+	// overallocate because we may have more than one conns per peer
+	selected := make([]network.Conn, 0, 2*target)
+	seen := make(map[network.Conn]struct{})
+
+	// first select connections that haven't seen a stream since some time.
+	for _, inf := range candidates {
+		if target <= 0 {
+			break
+		}
+
+		// lock this to protect from concurrent modifications from connect/disconnect events
+		s := cm.segments.get(inf.id)
+		s.Lock()
+
+		for c, info := range inf.conns {
+			if !info.lastStreamOpen.IsZero() && time.Since(info.lastStreamOpen) > maxStreamOpenDuration {
+				selected = append(selected, c)
+				target--
+				seen[c] = struct{}{}
```

Comment:
we should only do this if there are no active streams.

Reply:
And because it might be expensive to count on demand, we can keep a running count using the StreamOpened/StreamClosed events.

Reply:
@vyzo 

The current heuristic is:-
"Cleanup connections that vane't seen a stream since 10 Minutes (but could have open streams older than that)".

Are you saying that we should clean-up streams that:
"Don't have any open streams at all inspite of being older than 10 minutes" ?

 Any reason to prefer the latter ?

Reply:
That heuristic is incorrect, it will break applications with long lived stream.
Imagine a pubsub client, who simply connects and opens the pubsub streams and does nothing else afterwards.
If we reap the connection because there are no new streams, we just broke the app.

So yeah, we should only close connections if they have been idle for 10min _without_ any streams present.

Reply:
make sense.